library/train_util.py

```diff
@@ -4705,112 +4710,141 @@ def sample_images_common(
         requires_safety_checker=False,
         clip_skip=args.clip_skip,
     )
-    pipeline.to(device)
-
+    pipeline.to(distributed_state.device)
     save_dir = args.output_dir + "/sample"
     os.makedirs(save_dir, exist_ok=True)
-
+    temp_prompts = []
+    for i, prompt_dict in enumerate(prompts):
+        if isinstance(prompt_dict, str):
+            prompt_dict = line_to_prompt_dict(prompt_dict)
+        assert isinstance(prompt_dict, dict)
+        # Adds an enumerator to the dict based on prompt position. Used later to name image files. Also cleanup of extra data in original prompt dict.
+        temp_dict: dict = {}
+        temp_dict["negative_prompt"] = prompt_dict.get("negative_prompt")
+        temp_dict["sample_steps"] = prompt_dict.get("sample_steps", 30)
+        temp_dict["width"] = prompt_dict.get("width", 512)
+        temp_dict["height"] = prompt_dict.get("height", 512)
+        temp_dict["scale"] = prompt_dict.get("scale", 7.5)
+        temp_dict["seed"] = prompt_dict.get("seed")
+        temp_dict["controlnet_image"] = prompt_dict.get("controlnet_image")
+        temp_dict["prompt"]: str = prompt_dict.get("prompt", "")
+        temp_dict["sample_sampler"]: str = prompt_dict.get("sample_sampler", args.sample_sampler)
+        temp_dict["enum"] = i
+        temp_prompts.append(temp_dict)
+    prompts = temp_prompts
+    temp_prompts = []
+    num_of_processes = distributed_state.num_processes
+    # Creating list with N elements, where each element is a list of prompt_dicts, and N is the number of processess available (number of devices available)
+    # prompt_dicts are assigned to lists based on order of processes, to attempt to time the image creation time to match enum order. Probably only works when steps and sampler are identical.
+    for i in range(num_of_processes):
+        temp_prompts.append([])
+    for i in range(len(prompts)):
+        temp_prompts[i%num_of_processes].append(prompts[i])
+    prompts=temp_prompts
+    del temp_prompts
```

Comment:
This would be better refactored to a separate function too.

Reply:
refactored, not yet tested

Reply:
> Thank you for this PR.
> 
> I'm not sure why `temp_dict` is needed. I think the original implementation (assign variables from `prompt_dict` for each prompt) is fine. Is there any particular reason for this?

Strictly speaking, the original code sort of work well enough, if you don't care about the order of the sample images. The modifications here was mainly originally to add the "enum" to enumerate the sample prompts based on the order provided in the sample file. The original code based the index numbering on the list order that is passed to the image generation pipeline, which works elegantly when there is only one pipeline.

However, let's say there is a list of [1,2,3,4,5,6] prompt dicts that's [passed to the distributed GPU processes](https://huggingface.co/docs/accelerate/usage_guides/distributed_inference), the list will be split in the following pattern: 

- 2 GPUs [1,2,3] [4,5,6];
- 3 GPUs [1,2] [3,4] [5,6];
- 4 GPUs [1,2] [3,4] [5] [6];
- and so forth

Using the original code, what would happen is that in the case of 2 GPUs, 1 and 4 would be generated first, and indexed as the first sample prompt, 2 and 5 next, and so forth.

As I judged that the order of the sample images is somewhat important in order to correctly recognize what the effects are on the sample prompts provided, which is why I ran through the prompt_dict to add the enumerator, and decided to "clean-up" the list to only have the necessary info. (the orginal prompt_dicts still contained full lists of the prompt subsets).

That is also why I specifically set up to assign the prompt dicts into separate sub-lists matching the number of GPUs later on in the code, to get the image generation time to match the the order of the prompts as closely as possible.

However, later testing on Kaggle shows that even setting the seed from random to specific was sufficient to add enough delay to throw off the order. Probably will still need the accelerator.wait_for_everyone() at the end of each image generation in order to ensure timing matches perfectly.

Reply:
> That makes sense. However, I think the following part does it:
> 
> ```
>     per_process_prompts = [[] for i in range(num_of_processes)]
>     for i, prompt in enumerate(prompts):
>         per_process_prompts[i % num_of_processes].append(prompt)
> ```
> 
> I'm still not sure why `tmp_dict` is needed.

Hrm. You are right, should be possible to eliminate that section of code if we drop the clean-up effect, and add a enumerator dict in the loop before we pass it to the different processes.

Lemme poke and clean up that part first

Reply:
I've rewritten the code to simplify it, and tested it as working on Google Colab, managed to still cleanup the prompts by popping out the unused "subsets" in the dict. Don't think it would have much effect, but testing in some enviroments show taking the effort to remove the unused elements actually speeds up execution.

Also still pondering about the output timestamp. While I don't think it's ideal to sacrifice processing performance to ensure files are created in the exact order of the prompt, perhaps the issue could be resolved by changing the naming scheme of the sample images?

Instead of:
```
img_filename = (
    f"{'' if args.output_name is None else args.output_name + '_'}{ts_str}_{num_suffix}_{i:02d}{seed_suffix}.png"
)
```
Changing it to 
```
img_filename = (
    f"{'' if args.output_name is None else args.output_name + '_'}{num_suffix}_{i:02d}_{ts_str}{seed_suffix}.png"
)
```
would have the images sorted by filename in the same order as the sample prompts. Currently sorting by created time or name has the same effect